crates/services/txpool_v2/src/spent_inputs.rs

```diff
@@ -0,0 +1,283 @@
+        Input,
+        TxId,
+        UtxoId,
+    },
+    fuel_types::Nonce,
+};
+use lru::LruCache;
+use std::{
+    collections::HashMap,
+    num::NonZeroUsize,
+};
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+enum InputKey {
+    Utxo(UtxoId),
+    Message(Nonce),
+}
+
+/// Keeps track of spent inputs.
+pub struct SpentInputs {
+    /// Use LRU to support automatic clean up of old entries.
+    spent_inputs: LruCache<InputKey, ()>,
+    /// When it is unclear whether an input has been spent, we want to store which
+    /// transaction spent it. Later, this information can be used to unspent
+    /// or fully spend the input.
+    spender_of_inputs: HashMap<TxId, Vec<InputKey>>,
+}
+
+impl SpentInputs {
+    pub fn new(capacity: NonZeroUsize) -> Self {
+        Self {
+            spent_inputs: LruCache::new(capacity),
+            spender_of_inputs: HashMap::new(),
+        }
+    }
+
+    pub fn maybe_spend_inputs(&mut self, tx_id: TxId, inputs: &[Input]) {
+        let inputs = inputs
+            .iter()
+            .filter_map(|input| {
+                if input.is_coin() {
+                    input.utxo_id().cloned().map(InputKey::Utxo)
+                } else if input.is_message() {
+                    input.nonce().cloned().map(InputKey::Message)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for input in inputs.iter() {
+            self.spent_inputs.put(*input, ());
+        }
+        self.spender_of_inputs.insert(tx_id, inputs);
+    }
+
+    pub fn spend_inputs(&mut self, tx_id: TxId, inputs: &[Input]) {
+        let inputs = inputs.iter().filter_map(|input| {
+            if input.is_coin() {
+                input.utxo_id().cloned().map(InputKey::Utxo)
+            } else if input.is_message() {
+                input.nonce().cloned().map(InputKey::Message)
+            } else {
+                None
+            }
+        });
+
+        for input in inputs {
+            self.spent_inputs.put(input, ());
+        }
+        self.spend_inputs_by_tx_id(tx_id);
+    }
+
+    pub fn spend_inputs_by_tx_id(&mut self, tx_id: TxId) {
+        let inputs = self.spender_of_inputs.remove(&tx_id);
+
+        if let Some(inputs) = inputs {
+            for input in inputs {
+                self.spent_inputs.put(input, ());
+            }
+        }
+    }
+
+    pub fn unspend_inputs(&mut self, tx_id: TxId) {
+        let inputs = self.spender_of_inputs.remove(&tx_id);
+
+        if let Some(inputs) = inputs {
+            for input in inputs {
+                self.spent_inputs.pop(&input);
+            }
+        }
+    }
+
+    pub fn is_spent_utxo(&self, input: &UtxoId) -> bool {
+        self.spent_inputs.contains(&InputKey::Utxo(*input))
+    }
+
+    pub fn is_spent_message(&self, input: &Nonce) -> bool {
+        self.spent_inputs.contains(&InputKey::Message(*input))
+    }
+}
+
+#[cfg(test)]
+#[allow(non_snake_case)]
+mod tests {
```

Comment:
Is it possible to move these tests to the service level? I don't think we should have tests on the internal details of the service.